### PR TITLE
MC-07 Dashboard data module

### PR DESCRIPTION
**Pull request by bugbenderx**

### CODE REVIEW

File: data/src/test/java/com/example/data/dashboard/BaseDashboardRepositoryTest.kt [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+        )
+    }
+
+
+    @Test
+    fun testInvalidCurrencyPairAndNoInternet(): Unit = runBlocking {
+        favoriteCurrenciesCacheDataSource.hasCache(actualData = false)
+        currencyConverterCloudDataSource.returnFailure(noConnection = true)
+
+        val result = repository.dashboards()
```
Comment (JohnnySC) [inline data/src/test/java/com/example/data/dashboard/BaseDashboardRepositoryTest.kt]: val actual =

File: data/src/test/java/com/example/data/dashboard/BaseDashboardRepositoryTest.kt [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+                    DashboardItem.Base("C", "D", 1.3)
+                )
+            ), result
+        )
+    }
+
+
+    @Test
+    fun testInvalidCurrencyPairAndNoInternet(): Unit = runBlocking {
+        favoriteCurrenciesCacheDataSource.hasCache(actualData = false)
```
Comment (JohnnySC) [inline data/src/test/java/com/example/data/dashboard/BaseDashboardRepositoryTest.kt]: hasValidCache

File: data/src/test/java/com/example/data/dashboard/BaseDashboardRepositoryTest.kt [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+                    DashboardItem.Base("C", "D", 1.3)
+                )
+            ), result
+        )
+    }
+
+
+    @Test
+    fun testUserHasInValidCurrencyPairs(): Unit = runBlocking {
+        favoriteCurrenciesCacheDataSource.hasCache(actualData = false)
```
Comment (JohnnySC) [inline data/src/test/java/com/example/data/dashboard/BaseDashboardRepositoryTest.kt]: hasInvalidCache

File: data/src/main/java/com/example/data/dashboard/DashboardItemMapper.kt [hunk 4]
```diff
@@ -0,0 +1,46 @@
+package com.example.data.dashboard
+
+import com.example.data.dashboard.cache.CurrencyPairEntity
+import com.example.data.dashboard.cache.CurrentDate
+import com.example.data.dashboard.cache.FavoriteCurrenciesCacheDataSource
+import com.example.data.dashboard.cloud.CurrencyConverterCloudDataSource
+import com.example.domain.dashboard.DashboardItem
+
+interface DashboardItemMapper {
```
Comment (JohnnySC) [inline data/src/main/java/com/example/data/dashboard/DashboardItemMapper.kt]: Unit Test

File: data/src/main/java/com/example/data/dashboard/cache/CurrencyPairEntity.kt [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    val fromCurrency: String,
+    @ColumnInfo("to")
+    val toCurrency: String,
+    @ColumnInfo("rates")
+    val rates: Double = -1.0,
+    @ColumnInfo("data")
+    val date: String = ""
+) {
+
+    fun isInvalidRate(currentDate: CurrentDate): Boolean {
```
Comment (JohnnySC) [inline data/src/main/java/com/example/data/dashboard/cache/CurrencyPairEntity.kt:18]: unit test for this method

File: data/src/main/java/com/example/data/dashboard/BaseDashboardRepository.kt [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+        val favoriteCurrencies = favoriteCacheDataSource.favoriteCurrencies()
+
+        return if (favoriteCurrencies.isEmpty()) {
+            DashboardResult.Empty
+        } else {
+            try {
+                val listOfItems: List<DashboardItem> = mapper.map(favoriteCurrencies)
+                DashboardResult.Success(listOfItems = listOfItems)
+            } catch (e: Exception) {
+                val message = if (e is UnknownHostException) {
```
Comment (JohnnySC) [inline data/src/main/java/com/example/data/dashboard/BaseDashboardRepository.kt]: interface HandleError {

fun handle(error: Exception) : String


class Base (private val provideResources : ProvideResources) : HandleError {

override fun handle(error: Exception) : String {
if (e is UnknownHostException) {
 provideResources.noInternetConnectionMessage()
                } else {
                    provideResources.serviceUnavailableMessage()
                }

**Review by JohnnySC — CHANGES_REQUESTED** [review]

File: data/src/test/java/com/example/data/dashboard/BaseDashboardRepositoryTest.kt [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+
+    @Before
+    fun setUp() {
+        favoriteCurrenciesCacheDataSource = FakeFavoriteCurrenciesCacheDataSource()
+        currencyConverterCloudDataSource = FakeCurrencyConverterCloudDataSource()
+        handleError = FakeHandleError()
+        currentDate = FakeCurrentDate("1/1/2024")
+        repository = BaseDashboardRepository(
+            favoriteCacheDataSource = favoriteCurrenciesCacheDataSource,
+            mapper = DashboardItemMapper.Base(
```
Comment (JohnnySC) [inline data/src/test/java/com/example/data/dashboard/BaseDashboardRepositoryTest.kt]: fakeMapper

File: data/src/test/java/com/example/data/dashboard/cache/CurrencyPairEntityTest.kt [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+
+import com.example.data.dashboard.core.FakeCurrentDate
+import org.junit.Assert.assertEquals
+import org.junit.Test
+
+class CurrencyPairEntityTest {
+
+    private val currentDate = FakeCurrentDate("28/02/2024")
+
+    @Test
```
Comment (JohnnySC) [inline data/src/test/java/com/example/data/dashboard/cache/CurrencyPairEntityTest.kt:11]: 4th test

File: data/src/test/java/com/example/data/dashboard/BaseDashboardRepositoryTest.kt [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+
+        assertEquals(
+            DashboardResult.Error(message = "serviceUnavailable"), actualResult
+        )
+    }
+}
+
+private class FakeHandleError : HandleError {
+
+    override fun handle(e: Exception) =
```
Comment (JohnnySC) [inline data/src/test/java/com/example/data/dashboard/BaseDashboardRepositoryTest.kt]: handle(e:Exception): String = e.javaclass.name

File: data/src/test/java/com/example/data/dashboard/core/FakeCurrencyConverterCloudDataSource.kt [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+
+import com.example.data.dashboard.cloud.CurrencyConverterCloudDataSource
+import java.net.UnknownHostException
+
+internal class FakeCurrencyConverterCloudDataSource : CurrencyConverterCloudDataSource {
+
+    private var returnSuccess = true
+    private var noInternet = true
+
+    override suspend fun exchangeRate(from: String, to: String): Double {
```
Comment (JohnnySC) [inline data/src/test/java/com/example/data/dashboard/core/FakeCurrencyConverterCloudDataSource.kt:7]: return 1.1

File: data/src/test/java/com/example/data/dashboard/core/FakeCurrencyConverterCloudDataSource.kt [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+        return if (returnSuccess) {
+            99.9
+        } else {
+            throw if (noInternet) UnknownHostException() else RuntimeException()
```
Comment (JohnnySC) [inline data/src/test/java/com/example/data/dashboard/core/FakeCurrencyConverterCloudDataSource.kt]: throw 1 exception

**Review by JohnnySC — CHANGES_REQUESTED** [review]

File: data/src/test/java/com/example/data/dashboard/cache/CurrencyPairEntityTest.kt (hunk 8 above)
Comment (k0dm) [reply data/src/test/java/com/example/data/dashboard/cache/CurrencyPairEntityTest.kt:11]: no need

File: data/src/main/java/com/example/data/dashboard/cloud/CurrencyConverterCloudDataSource.kt [hunk 12]
```diff
@@ -0,0 +1,12 @@
+package com.example.data.dashboard.cloud
+
+interface CurrencyConverterCloudDataSource {
```
Comment (JohnnySC) [inline data/src/main/java/com/example/data/dashboard/cloud/CurrencyConverterCloudDataSource.kt:3]: test check not null

File: data/src/test/java/com/example/data/dashboard/cache/CurrencyPairEntityTest.kt [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+    fun testInvalidDate() {
+        val currencyPairEntity = CurrencyPairEntity("1", "2", 1.0, "27/02/2024")
+        assertEquals(currencyPairEntity.isInvalidRate(currentDate), true)
+    }
+
+    @Test
+    fun testInvalidRateAndDate() {
+        val currencyPairEntity = CurrencyPairEntity("1", "2", -1.0, "27/02/2024")
+        assertEquals(currencyPairEntity.isInvalidRate(currentDate), true)
```
Comment (JohnnySC) [inline data/src/test/java/com/example/data/dashboard/cache/CurrencyPairEntityTest.kt]: assertEquals(actual, expected)

File: data/src/main/java/com/example/data/dashboard/HandleError.kt [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+import com.example.data.core.ProvideResources
+import java.net.UnknownHostException
+
+interface HandleError {
+
+    fun handle(e: Exception): String
+
+    class Base(private val provideResources: ProvideResources) : HandleError {
+
+        override fun handle(e: Exception) = if (e is UnknownHostException) {
```
Comment (JohnnySC) [inline data/src/main/java/com/example/data/dashboard/HandleError.kt]: with(provideResources)

File: data/src/main/java/com/example/data/dashboard/DashboardItemsDatasource.kt [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+
+import com.example.data.dashboard.cache.CurrencyPairEntity
+import com.example.data.dashboard.cache.CurrentDate
+import com.example.data.dashboard.cache.FavoriteCurrenciesCacheDataSource
+import com.example.data.dashboard.cloud.CurrencyConverterCloudDataSource
+import com.example.domain.dashboard.DashboardItem
+
+interface DashboardItemsDatasource {
+
+    suspend fun map(favoriteCurrencies: List<CurrencyPairEntity>): List<DashboardItem>
```
Comment (JohnnySC) [inline data/src/main/java/com/example/data/dashboard/DashboardItemsDatasource.kt]: fun dashboardItems(

**Review by JohnnySC — CHANGES_REQUESTED** [review]

File: data/src/main/java/com/example/data/dashboard/cache/CurrencyPairEntity.kt [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+        return rates == -1.0 || date != currentDate.date()
+    }
+}
+
+interface CurrentDate {
+
+    fun date(): String
+
+    class Base() : CurrentDate {
```
Comment (JohnnySC) [inline data/src/main/java/com/example/data/dashboard/cache/CurrencyPairEntity.kt]: class Base(private val date: Date = LocalDateTime.now()
private val format : String = "yyyy/MM/dd")

**Review by JohnnySC — CHANGES_REQUESTED** [review]

File: data/src/test/java/com/example/data/dashboard/cache/CurrentDateTest.kt [hunk 17]
```diff
@@ -0,0 +1,19 @@
+package com.example.data.dashboard.cache
+
+import org.junit.Assert.assertEquals
+import org.junit.Test
+
+class CurrentDateTest {
+
+    @Test
+    fun test() {
+        var currentDate = CurrentDate.Base(format = "yyyy")
```
Comment (JohnnySC) [inline data/src/test/java/com/example/data/dashboard/cache/CurrentDateTest.kt]: val actual = CurrentDate.Base(LocalDateTime.of(2024, 2, 28, 21, 15, 0), format="dd/mm/yyyy HH:mm:ss").date()
val expected = "28/2/2024 21:15:00"

**Review by JohnnySC — CHANGES_REQUESTED** [review]

**Review by JohnnySC — APPROVED** [review]

**Pull request merged**